README.md
[![CI](https://github.com/ethanbergstrom/Cobalt/actions/workflows/CI.yml/badge.svg)](https://github.com/ethanbergstrom/Cobalt/actions/workflows/CI.yml)

# Cobalt
Cobalt is a simple PowerShell Crescendo wrapper for WinGet

## Requirements
In addition to PowerShell 5.1+ and an Internet connection on a Windows machine, WinGet must also be installed. Microsoft recommends installing WinGet from the Windows Store as part of the [App Installer](https://www.microsoft.com/en-us/p/app-installer/9nblggh4nns1?activetab=pivot:overviewtab) package.

## Install Cobalt
```PowerShell
Install-Module Cobalt -Force
```

## Sample usages
### Search for a package
```PowerShell
Find-WinGetPackage -ID openJS.nodejs

Find-WinGetPackage -ID Mozilla.Firefox -Exact
```

### Get a package's detailed information from the repository
```PowerShell
Get-WinGetPackageInfo -ID CPUID.CPU-Z -Version 1.95

Find-WinGetPackage -ID Mozilla.Firefox -Exact | Get-WinGetPackageInfo
```

### Get all available versions of a package
```PowerShell
Get-WinGetPackageInfo -ID CPUID.CPU-Z -Versions

Find-WinGetPackage -ID Mozilla.Firefox -Exact | Get-WinGetPackageInfo -Versions
```

### Install a package
```PowerShell
Find-WinGetPackage OpenJS.NodeJS -Exact | Install-WinGetPackage

Install-WinGetPackage 7zip.7zip
```

### Install a list of packages
```PowerShell
@('CPUID.CPU-Z','7zip.7zip') | ForEach-Object { Install-WinGetPackage $_ }
```

### Install a specific version of a package
```PowerShell
Install-WinGetPackage CPUID.CPU-Z -Version 1.95
```

### Install multiple packages with specific versions
```PowerShell
@(
    @{
        id = 'CPUID.CPU-Z'
        version = '1.95'
    },
    @{
        id = 'putty.putty'
        version = '0.74'
    }
) | Install-WinGetPackage
```

### Get list of installed packages
```PowerShell
Get-WinGetPackage nodejs
```

### Get list of installed packages that can be upgraded
```PowerShell
Get-WinGetPackageUpdate
```

### Upgrade a package
```PowerShell
Update-WinGetPackage CPUID.CPU-Z
```

### Upgrade a list of packages
```PowerShell
@('CPUID.CPU-Z','7zip.7zip') | ForEach-Object { Update-WinGetPackage -ID $_ }
```

### Upgrade all packages
> :warning: **Use at your own risk!** WinGet will try to upgrade all layered software it finds, may not always succeed, and may upgrade software you don't want upgraded.
```PowerShell
Update-WinGetPackage -All
```

### Uninstall a package
```PowerShell
Get-WinGetPackage nodejs | Uninstall-WinGetPackage

Uninstall-WinGetPackage 7zip.7zip
```

### Manage package sources
```PowerShell
Register-WinGetSource privateRepo -Argument 'https://somewhere/out/there/api/v2/'
Find-WinGetPackage nodejs -Source privateRepo -Exact | Install-WinGetPackage
Unregister-WinGetSource privateRepo
```

Cobalt integrates with WinGet.exe to manage and store source information

## Known Issues
### Stability
WinGet's behavior and APIs are still very unstable. Do not be surprised if this module stops working with newer versions of WinGet.

### Retrieving package upgrade list hangs on first use
Due to [a bug](https://github.com/microsoft/winget-cli/issues/1869) that [is resolved](https://github.com/microsoft/winget-cli/pull/1874) in WinGet v1.3 preview releases, if `Get-WinGetPackageUpdate` is ran with WinGet v1.2.x or below without having first accepted source license agreements, the cmdlet will hang indefinitely due to source agreements not having been accepted.

Available workarounds include:
* Running another cmdlet that invokes correctly-behaving WinGet behavior (ex: `Get-WinGetPackage`)
* Manually accepting the source agreement via the WinGet CLI

After WinGet v1.3 is generally available with the bug fixed, a new version of this module will be released to resolve this issue.

## Legal and Licensing
Cobalt is licensed under the [MIT license](./LICENSE.txt).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2023-02-05
### Changed
- Upgraded to PowerShell Crescendo 1.1 Preview 1 for compiling the module. No functional changes expected.

## [0.3.3] - 2023-01-29
### Fixed
- Hard-code the version of WinGet non-EN-US localization resources to work around microsoft/winget-cli#2783

## [0.3.2] - 2023-01-21
### Fixed
- Package version information for packages with long names correctly displayed in non-UTF-8 encoded output terminals

## [0.3.1] - 2022-05-13
### Fixed
- Package upgrade list functionality now correctly supports non-EN-US localities

## [0.3.0] - 2022-05-12
### Added
- Ability to return a list of packages that qualify for updates
### Changed
- `Get-WinGetPackage` now returns both installed and available version information

## [0.2.1] - 2022-03-12
### Fixed
- Package uninstallation error handling should now correctly catch failures

## [0.2.0] - 2022-03-12
### Added
- Ability to retrieve package metadata and versions
- Uninstall failure error handling
### Fixed
- Package installation error handling should now correctly catch failures with non-US English languages

## [0.1.0] - 2022-02-06
### Added
- Upgrade functionality

## [0.0.11] - 2022-01-22
### Changed
- Error output is more targeted to only what failed

## [0.0.10] - 2021-12-26
### Fixed
- Correctly source display language information
- Force console encoding in automated tests

## [0.0.9] - 2021-12-21
### Fixed
- Handle non-US English locales correctly

## [0.0.8] - 2021-12-21
### Fixed
- More dynamic locale-based column parsing

## [0.0.7] - 2021-12-04
### Fixed
- Even more string parsing corrections

## [0.0.6] - 2021-12-04
### Fixed
- Yet more string parsing corrections

## [0.0.5] - 2021-12-04
### Fixed
- Additional string parsing/cleaning corrections

## [0.0.4] - 2021-12-04
### Fixed
- Improved and consolidated string parsing/cleaning

## [0.0.3] - 2021-12-04
### Fixed
- Correctly order output package attributes

## [0.0.2] - 2021-12-04
### Fixed
- Correctly handle output when no results are found with `list` or `search` commands

## [0.0.1] - 2021-12-02
- Initial release


build.ps1
. (Join-Path -Path src -ChildPath Cobalt.ps1 -Resolve)

$commandArray = @()

$commands | ForEach-Object {
	$Noun = $_.Noun
	# Inherit noun-level attributes (if they exist) for all commands
	# If no noun-level original command elements or parameters exist, return an empty array for easy merging later
	$NounOriginalCommandElements = $_.OriginalCommandElements ?? @()
	$NounParameters = $_.Parameters ?? @()
	# Output handlers work differently - they will supercede each other, instead of being merged.
	$NounOutputHandlers = $_.OutputHandlers
	$NounDefaultParameterSetName = $_.DefaultParameterSetName
	$_.Verbs | ForEach-Object {
		# Same logic as nouns - prepare verb-level original command elements and parameters for merging, but not output handlers
		$VerbOriginalCommandElements = $_.OriginalCommandElements ?? @()
		$VerbParameters = $_.Parameters ?? @()
		$VerbOutputHandlers = $_.OutputHandlers
		$Description = $_.Description
		$VerbDefaultParameterSetName = $_.DefaultParameterSetName
		$commandArray += $(New-CrescendoCommand -Verb $_.Verb -Noun $Noun -OriginalName $BaseOriginalName | ForEach-Object {
			# Marge command elements in order of noun-level first, then verb-level, then generic
			$_.OriginalCommandElements = ($NounOriginalCommandElements + $VerbOriginalCommandElements + $BaseOriginalCommandElements)
			$_.Description = $Description
			# Merge parameters in order of noun-level, then verb-level, then generic
			$_.Parameters = ($NounParameters + $VerbParameters + $BaseParameters)
			# Prefer verb-level default parameter set name first, then noun-level, then generic
			$_.DefaultParameterSetName = ($VerbDefaultParameterSetName ?? $NounDefaultParameterSetName) ?? $BaseDefaultParameterSetName
			# Prefer verb-level handlers first, then noun-level, then generic
			$_.OutputHandlers = ($VerbOutputHandlers ?? $NounOutputHandlers) ?? $BaseOutputHandlers
			$_
		})
	}
}

$tempJson = (New-TemporaryFile).FullName
Export-CrescendoCommand -command $commandArray -fileName $tempJson -Force
Export-CrescendoModule -NoClobberManifest -ConfigurationFile $tempJson -ModuleName (Join-Path -Path src -ChildPath Cobalt.psm1) -Force


Install-WinGet.ps1
Install-Module NtObjectManager -Force
Import-Module appx -UseWindowsPowerShell -WarningAction SilentlyContinue

# GitHub release information
$appxPackageName = 'Microsoft.DesktopAppInstaller'
$msWinGetLatestReleaseURL = 'https://github.com/microsoft/winget-cli/releases/expanded_assets/v1.3.2691'
$msWinGetMSIXBundlePath = ".\$appxPackageName.msixbundle"
$msWinGetLicensePath = ".\$appxPackageName.license.xml"

# Workaround for no Microsoft Store on Windows Server - I dont know a great way to source this information dynamically
$architecture = 'x64'
$msStoreDownloadAPIURL = 'https://store.rg-adguard.net/api/GetFiles'
$msWinGetStoreURL = 'https://www.microsoft.com/en-us/p/app-installer/9nblggh4nns1'
$msVCLibPattern = "*Microsoft.VCLibs*UWPDesktop*$architecture*appx*"
$msVCLibDownloadPath = '.\Microsoft.VCLibs.UWPDesktop.appx'
$msUIXamlPattern = "*Microsoft.UI.Xaml*$architecture*appx*"
$msUIXamlDownloadPath = '.\Microsoft.UI.Xaml.appx'
$msWinGetExe = 'winget'
$wingetExecAliasPath = "C:\Windows\System32\$msWinGetExe.exe"

$msWinGetLatestRelease = Invoke-WebRequest -Uri $msWinGetLatestReleaseURL

# Download the latest MSIX bundle and matching license from GitHub
$msWinGetLatestRelease.links |
    Where-Object href -like '*msixbundle' |
        Select-Object -Property @{
            Name = 'URI';
            Expression = {$msWinGetLatestRelease.BaseResponse.headers.Server.Product.Name+$_.href}
        } | ForEach-Object {Invoke-WebRequest -Uri $_.URI -OutFile $msWinGetMSIXBundlePath}

# Hopefully this mitigates the sporadic authentication denied errors from GitHub's CDN
Start-Sleep -Seconds 10

$msWinGetLatestRelease.links |
    Where-Object href -Like '*License*xml' |
        Select-Object -Property @{
            Name = 'URI';
            Expression = {$msWinGetLatestRelease.BaseResponse.headers.Server.Product.Name+$_.href}
        } | ForEach-Object {Invoke-WebRequest -Uri $_.URI -OutFile $msWinGetLicensePath}

# Download the VC++ redistrubable for UWP apps from the Microsoft Store
(Invoke-WebRequest -Uri $msStoreDownloadAPIURL -Method Post -Form @{type='url'; url=$msWinGetStoreURL; ring='Retail'; lang='en-US'}).links |
    Where-Object OuterHTML -Like $msVCLibPattern |
        Sort-Object outerHTML -Descending |
            Select-Object -First 1 -ExpandProperty href |
                ForEach-Object {Invoke-WebRequest -Uri $_ -OutFile $msVCLibDownloadPath}

# Download the Windows UI redistrubable from the Microsoft Store
(Invoke-WebRequest -Uri $msStoreDownloadAPIURL -Method Post -Form @{type='url'; url=$msWinGetStoreURL; ring='Retail'; lang='en-US'}).links |
Where-Object OuterHTML -Like $msUIXamlPattern |
    Sort-Object outerHTML -Descending |
        Select-Object -First 1 -ExpandProperty href |
            ForEach-Object {Invoke-WebRequest -Uri $_ -OutFile $msUIXamlDownloadPath}

# Install the WinGet and it's VC++ .msix with the downloaded license file
Add-AppProvisionedPackage -Online -PackagePath $msWinGetMSIXBundlePath -DependencyPackagePath ($msVCLibDownloadPath,$msUIXamlDownloadPath) -LicensePath $msWinGetLicensePath

# Force the creation of the execution alias with NtObjectManager, since one isn't generated automatically in the current user session
$appxPackage = Get-AppxPackage Microsoft.DesktopAppInstaller
$wingetTarget = Join-Path -Path $appxPackage.InstallLocation -ChildPath ((Get-AppxPackageManifest $appxPackage).Package.Applications.Application | Where-Object Id -eq $msWinGetExe | Select-Object -ExpandProperty Executable)
NtObjectManager\Set-ExecutionAlias -Path $wingetExecAliasPath -PackageName ($appxPackage.PackageFamilyName) -EntryPoint "$($appxPackage.PackageFamilyName)!$msWinGetExe" -Target $wingetTarget -AppType Desktop -Version 3


src/Cobalt.ps1
$BaseOriginalName = 'WinGet'

$BaseOriginalCommandElements = @()

$BaseParameters = @()

$BaseOutputHandlers = @{
    ParameterSetName = 'Default'
    Handler = {
        param ( $output )
    }
}

$outputHanderHeader = {param ($output)}

$i18nHandlerHelper = {
    $language = (Get-UICulture).Name

    $languageData = $(
        $hash = @{}

        $(try {
            # We have to trim the leading BOM for .NET's XML parser to correctly read Microsoft's own files - go figure
            ([xml](((Invoke-WebRequest -Uri "https://raw.githubusercontent.com/microsoft/winget-cli/v1.3.2691/Localization/Resources/$language/winget.resw" -ErrorAction Stop ).Content -replace "\uFEFF", ""))).root.data
        } catch {
            # Fall back to English if a locale file doesn't exist
            (
                ('SearchName','Name'),
                ('SearchID','Id'),
                ('SearchVersion','Version'),
                ('AvailableHeader','Available'),
                ('SearchSource','Source'),
                ('ShowVersion','Version'),
                ('GetManifestResultVersionNotFound','No version found matching:'),
                ('InstallerFailedWithCode','Installer failed with exit code:'),
                ('UninstallFailedWithCode','Uninstall failed with exit code:'),
                ('AvailableUpgrades','upgrades available.')
            ) | ForEach-Object {[pscustomobject]@{name = $_[0]; value = $_[1]}}
        }) | ForEach-Object {
            # Convert the array into a hashtable
            $hash[$_.name] = $_.value
        }

        $hash
    )
}

$GetPackageOutputHandler = {
    $nameHeader = $output -Match "^$($languageData.SearchName)"

    if ($nameHeader) {

        $headerLine = $output.IndexOf(($nameHeader | Select-Object -First 1))

        if ($headerLine -ne -1) {
            $idIndex = $output[$headerLine].IndexOf(($languageData.SearchID))
            $versionIndex = $output[$headerLine].IndexOf(($languageData.SearchVersion))
            $availableIndex = $output[$headerLine].IndexOf(($languageData.AvailableHeader))
            $sourceIndex = $output[$headerLine].IndexOf(($languageData.SearchSource))

            # Stop gathering version data at the 'Available' column if it exists, if not continue on to the 'Source' column (if it exists)
            $versionEndIndex = $(
                if ($availableIndex -ne -1) {
                    $availableIndex
                } else {
                    $sourceIndex
                }
            )

            # Only attempt to parse output if it contains a 'version' column
            if ($versionIndex -ne -1) {
                # The -replace cleans up errant characters that come from WinGet's poor treatment of truncated columnar output
                ($output | Select-String -Pattern $languageData.AvailableUpgrades,'--include-unknown' -NotMatch) -replace '[^i\p{IsBasicLatin}]+',' ' | Select-Object -Skip ($headerLine+2) | ForEach-Object {
                    Remove-Variable -Name 'package' -ErrorAction SilentlyContinue

                    $package = [ordered]@{
                        ID = $_.SubString($idIndex,$versionIndex-$idIndex).Trim()
                    }

                    if ($package) {
                        # I'm so sorry, blame WinGet
                        # If neither the 'Available' or 'Source' column exist, gather version data to the end of the string
                        $package.Version = $(
                            if ($versionEndIndex -ne -1) {
                                $_.SubString($versionIndex,$versionEndIndex-$versionIndex)
                            } else {
                                $_.SubString($versionIndex)
                            }
                        ).Trim() -replace '[^\.\d]'

                        # Only attempt to add 'Available Version' data if the column exists
                        if ($availableIndex -ne -1) {
                            $package.Available = $(
                                if ($sourceIndex -ne -1) {
                                    $_.SubString($availableIndex,$sourceIndex-$availableIndex)
                                } else {
                                    $_.SubString($availableIndex)
                                }
                            ).Trim() -replace '[^\.\d]'
                        }

                        # If the 'Source' column was included in the output, include it in our output, too
                        if (($sourceIndex -ne -1) -And ($_.Length -ge $sourceIndex)) {
                            $package.Source = $_.SubString($sourceIndex).Trim() -split ' ' | Select-Object -Last 1
                        }

                        [pscustomobject]$package
                    }
                }
            }
        }
    }
}

$InstallPackageOutputHandler = {
    if ($output) {
        if ($output -match $languageData.InstallerFailedWithCode) {
            # Only show output that matches or comes after the 'failed' keyword
            Write-Error ($output[$output.IndexOf($($output -match $languageData.InstallerFailedWithCode | Select-Object -First 1))..($output.Length-1)] -join "`r`n")
        } else {
            $output | ForEach-Object {
                if ($_ -match 'Found .+ \[(?<id>[\S]+)\] Version (?<version>[\S]+)' -and $Matches.id -and $Matches.version) {
                    [pscustomobject]@{
                        ID = $Matches.id
                        Version = $Matches.version
                    }
                }
            }
        }
    }
}

$UnInstallPackageOutputHandler = {
    if ($output) {
        if ($output -match $languageData.UninstallFailedWithCode) {
            # Only show output that matches or comes after the 'failed' keyword
            Write-Error ($output[$output.IndexOf($($output -match $languageData.UninstallFailedWithCode | Select-Object -First 1))..($output.Length-1)] -join "`r`n")
        }
    }
}

$PackageInfoVersionOutputHandler = {
    if ($output) {
        if ($output | Select-String -Pattern $languageData.GetManifestResultVersionNotFound) {
            # Only show output that matches or comes after the 'failed' keyword
            Write-Error ($output[$output.IndexOf($($output | Select-String -Pattern $languageData.GetManifestResultVersionNotFound | Select-Object -First 1))..($output.Length-1)] -join "`r`n")
        } else {
            $versionHeader = $output -Match "^$($languageData.ShowVersion)"

            if ($versionHeader) {

                $headerLine = $output.IndexOf(($versionHeader | Select-Object -First 1))

                if ($headerLine -ne -1) {
                    $output | Select-Object -Skip ($headerLine+2)
                }
            }
        }
    }
}

$RenderedGetPackageOutputHandler = [scriptblock]::Create($outputHanderHeader.ToString() + $i18nHandlerHelper.ToString() + $GetPackageOutputHandler.ToString())
$RenderedInstallPackageOutputHandler = [scriptblock]::Create($outputHanderHeader.ToString() + $i18nHandlerHelper.ToString() + $InstallPackageOutputHandler.ToString())
$RenderedUnInstallPackageOutputHandler = [scriptblock]::Create($outputHanderHeader.ToString() + $i18nHandlerHelper.ToString() + $UnInstallPackageOutputHandler.ToString())
$RenderedPackageInfoVersionOutputHandler = [scriptblock]::Create($outputHanderHeader.ToString() + $i18nHandlerHelper.ToString() + $PackageInfoVersionOutputHandler.ToString())

# The general structure of this hashtable is to define noun-level attributes, which are -probably- common across all commands for the same noun, but still allow for customization at more specific verb-level defition for that noun.
# The following three command attributes have the following order of precedence:
# 	OriginalCommandElements will be MERGED in the order of Noun + Verb + Base
#		Example: Noun WinGetSource's element 'source', Verb Register's element 'add', and Base elements are merged to become 'WinGet source add --limit-output --yes'
# 	Parameters will be MERGED in the order of Noun + Verb + Base
#		Example: Noun WinGetPackage's parameters for package name and version and Verb Install's parameter specifying source information are merged to become '<packageName> --version=<packageVersion> --source=<packageSource>'.
#			These are then appended to the merged original command elements, to create 'WinGet install <packageName> --version=<packageVersion> --source=<packageSource> --limit-output --yes'
# 	OutputHandler sets will SUPERCEDE each other in the order of: Verb -beats-> Noun -beats-> Base. This allows reusability of PowerShell parsing code.
#		Example: Noun WinGetPackage has inline output handler PowerShell code with complex regex that works for both Install-WinGetPackage and Uninstall-WinGetPackage, but Get-WinGetPackage's native output uses simple vertical bar delimiters.
#		Example 2: The native commands for Register-WinGetSource and Unregister-WinGetSource don't return any output, and until Crescendo supports error handling by exit codes, a base required default output handler that doesn't do anything can be defined and reused in multiple places.
$Commands = @(
    @{
        Noun = 'WinGetSource'
        OriginalCommandElements = @('source')
        Verbs = @(
            @{
                Verb = 'Get'
                Description = 'Return WinGet package sources'
                OriginalCommandElements = @('export')
                Parameters = @(
                    @{
                        Name = 'Name'
                        ParameterType = 'string'
                        Description = 'Source Name'
                        OriginalName = '--name='
                        NoGap = $true
                    }
                )
                OutputHandlers = @{
                    ParameterSetName = 'Default'
                    Handler = {
                        param ($output)
                        if ($output) {
                            $output | ConvertFrom-Json
                        }
                    }
                }
            },
            @{
                Verb = 'Register'
                Description = 'Register a new WinGet package source'
                OriginalCommandElements = @('add')
                Parameters = @(
                    @{
                        Name = 'Name'
                        ParameterType = 'string'
                        Description = 'Source Name'
                        OriginalName = '--name='
                        NoGap = $true
                        Mandatory = $true
                    },
                    @{
                        Name = 'Argument'
                        OriginalName = '--arg='
                        ParameterType = 'string'
                        Description = 'Source Argument'
                        NoGap = $true
                        Mandatory = $true
                    }
                )
                OutputHandlers = @{
                    ParameterSetName = 'Default'
                    Handler = {
                        param ($output)
                        if ($output) {
                            if ($output[-1] -ne 'Done') {
                                Write-Error ($output -join "`r`n")
                            }
                        }
                    }
                }
            },
            @{
                Verb = 'Unregister'
                Description = 'Unegister an existing WinGet package source'
                OriginalCommandElements = @('remove')
                Parameters = @(
                    @{
                        Name = 'Name'
                        ParameterType = 'string'
                        Description = 'Source Name'
                        OriginalName = '--name='
                        NoGap = $true
                        Mandatory = $true
                        ValueFromPipelineByPropertyName = $true
                    }
                )
                OutputHandlers = @{
                    ParameterSetName = 'Default'
                    Handler = {
                        param ($output)
                        if ($output) {
                            if ($output[-1] -match 'Did not find a source') {
                                Write-Error ($output -join "`r`n")
                            }
                        }
                    }
                }
            }
        )
    },
    @{
        Noun = 'WinGetPackage'
        Parameters = @(
            @{
                Name = 'ID'
                OriginalName = '--id='
                ParameterType = 'string'
                Description = 'Package ID'
                NoGap = $true
                ValueFromPipelineByPropertyName = $true
            },
            @{
                Name = 'Exact'
                OriginalName = '--exact'
                ParameterType = 'switch'
                Description = 'Search by exact package name'
            },
            @{
                Name = 'Source'
                OriginalName = '--source='
                ParameterType = 'string'
                Description = 'Package Source'
                NoGap = $true
                ValueFromPipelineByPropertyName = $true
            }
        )
        OutputHandlers = @{
            ParameterSetName = 'Default'
            Handler = $RenderedGetPackageOutputHandler
        }
        Verbs = @(
            @{
                Verb = 'Install'
                Description = 'Install a new package with WinGet'
                OriginalCommandElements = @('install','--accept-package-agreements','--accept-source-agreements','--silent')
                Parameters = @(
                    @{
                        Name = 'Version'
                        OriginalName = '--version='
                        ParameterType = 'string'
                        Description = 'Package Version'
                        NoGap = $true
                        ValueFromPipelineByPropertyName = $true
                    }
                )
                OutputHandlers = @{
                    ParameterSetName = 'Default'
                    Handler = $RenderedInstallPackageOutputHandler
                }
            },
            @{
                Verb = 'Get'
                Description = 'Get a list of installed WinGet packages'
                OriginalCommandElements = @('list','--accept-source-agreements')
            },
            @{
                Verb = 'Find'
                Description = 'Find a list of available WinGet packages'
                OriginalCommandElements = @('search','--accept-source-agreements')
            },
            @{
                Verb = 'Update'
                Description = 'Updates an installed package to the latest version'
                OriginalCommandElements = @('upgrade','--accept-source-agreements','--silent')
                Parameters = @(
                    @{
                        Name = 'All'
                        OriginalName = '--all'
                        ParameterType = 'switch'
                        Description = 'Upgrade all packages'
                    }
                )
                OutputHandlers = @{
                    ParameterSetName = 'Default'
                    Handler = $RenderedInstallPackageOutputHandler
                }
            },
            @{
                Verb = 'Uninstall'
                Description = 'Uninstall an existing package with WinGet'
                OriginalCommandElements = @('uninstall','--accept-source-agreements','--silent')
                OutputHandlers = @{
                    ParameterSetName = 'Default'
                    Handler = $RenderedUnInstallPackageOutputHandler
                }
            }
        )
    },
    @{
        Noun = 'WinGetPackageInfo'
        Verbs = @(
            @{
                Verb = 'Get'
                Description = 'Shows information on a specific WinGet package'
                OriginalCommandElements = @('show','--accept-source-agreements')
                DefaultParameterSetName = 'Default'
                Parameters = @(
                    @{
                        Name = 'ID'
                        OriginalName = '--id='
                        ParameterType = 'string'
                        Description = 'Package ID'
                        NoGap = $true
                        Mandatory = $true
                        ValueFromPipelineByPropertyName = $true
                        Position = 0
                        ParameterSetName = @('Default','Versions')
                    },
                    @{
                        Name = 'Exact'
                        OriginalName = '--exact'
                        ParameterType = 'switch'
                        Description = 'Search by exact package name'
                        ParameterSetName = @('Default','Versions')
                    },
                    @{
                        Name = 'Version'
                        OriginalName = '--version='
                        ParameterType = 'string'
                        Description = 'Package Version'
                        NoGap = $true
                        ValueFromPipelineByPropertyName = $true
                        ParameterSetName = @('Default','Versions')
                    },
                    @{
                        Name = 'Source'
                        OriginalName = '--source='
                        ParameterType = 'string'
                        Description = 'Package Source'
                        NoGap = $true
                        ValueFromPipelineByPropertyName = $true
                        ParameterSetName = @('Default','Versions')
                    },
                    @{
                        Name = 'Versions'
                        OriginalName = '--versions'
                        ParameterType = 'switch'
                        Description = 'Show available versions of the package'
                        ParameterSetName = 'Versions'
                    }
                )
                OutputHandlers = @(
                    @{
                        ParameterSetName = 'Default'
                        Handler = {
                            param ( $output )

                            $packageInfo = @{}

                            $output | Select-String -AllMatches -Pattern '^\s*([\w\s]+):\s(.+)$' | ForEach-Object -MemberName Matches | ForEach-Object{
                                $match = ($_.Groups | Select-Object -Skip 1).Value
                                $packageInfo.add($match[0],$match[1])
                            }

                            $packageInfo
                        }
                    },
                    @{
                        ParameterSetName = 'Versions'
                        Handler = $RenderedPackageInfoVersionOutputHandler
                    }
                )
            }
        )
    },
    @{
        Noun = 'WinGetPackageUpdate'
        OutputHandlers = @{
            ParameterSetName = 'Default'
            Handler = $RenderedGetPackageOutputHandler
        }
        Verbs = @(
            @{
                Verb = 'Get'
                Description = 'Get a list of installed WinGet packages'
                # Add this back in after WinGet 1.3 is released
                # https://github.com/microsoft/winget-cli/issues/1869
                # https://github.com/microsoft/winget-cli/pull/1874
                # OriginalCommandElements = @('upgrade','--accept-source-agreements')
                OriginalCommandElements = @('upgrade')
            }
        )
    }
)
